compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -109,6 +109,112 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                tcx.def_path_str(trait_pred.trait_ref.def_id)
+                    == tcx.def_path_str(into_iterator_trait)
+            } else {
+                false
+            }
+        }
+
+        fn is_iterator_predicate(predicate: ty::Predicate<'_>, tcx: TyCtxt<'_>) -> bool {
+            if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_pred)) =
+                predicate.kind().as_ref().skip_binder()
+                && let Some(iterator_trait) = tcx.get_diagnostic_item(sym::Iterator)
+            {
+                tcx.def_path_str(trait_pred.trait_ref.def_id) == tcx.def_path_str(iterator_trait)
+            } else {
+                false
+            }
+        }
+
+        match ty.kind() {
+            ty::Param(param) => {
+                let predicates = self.tcx.predicates_of(self.body_id);
+                let generics = self.tcx.generics_of(self.body_id);
+                let generic_param = generics.type_param(&param, self.tcx);
+                for clause in predicates.predicates.iter() {
+                    // Checking to see if this predicate sets an impl `IntoIterator` bound for `ty`
+                    if predicate_bounds_generic_param(
+                        clause.0.as_predicate(),
+                        generics,
+                        generic_param,
+                        self.tcx,
+                    ) && is_into_iterator_predicate(clause.0.as_predicate(), self.tcx)
+                    {
+                        for unsatisfied in unsatisfied_predicates.iter() {
+                            // The specified bounds state that the parameter implements `IntoIterator`
+                            // but it has called a method that requires it to implement `Iterator`
+                            if predicate_bounds_generic_param(
+                                unsatisfied.0,
+                                generics,
+                                generic_param,
+                                self.tcx,
+                            ) && is_iterator_predicate(unsatisfied.0, self.tcx)
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            ty::Alias(ty::AliasKind::Opaque, alias_ty) => {
+                let item_bounds = self.tcx.explicit_item_bounds(alias_ty.def_id);
+                let item_bounds = item_bounds.as_ref().skip_binder();
+                for clause in item_bounds.iter() {
+                    if is_into_iterator_predicate(clause.0.as_predicate(), self.tcx) {
+                        for unsatisfied in unsatisfied_predicates.iter() {
+                            if is_iterator_predicate(unsatisfied.0, self.tcx) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            _ => return false,
+        }
+        false
```

Comment:
I feel like this whole approach could be simplified by just checking if `{ty}: IntoIterator` holds. You can use `predicate_must_hold_modulo_regions` to do this.